Fix assigning to dynamically allocated structs

xbauch opened the pull request:
that is when
```
lhs :: [char]
rhs :: byte-extract-expr(struct, 0)
```
before we created a member expression at the offset and, in this case, assigned
to the first member of the struct. Now we assign the whole struct.

A test is included that was previously failing in the unwinding assertion (CBMC
did not see that the next pointer is eventually NULL). Which is issue #5002.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on regression/cbmc/dynamic_struct/main.c:
```diff
@@ -0,0 +1,55 @@
+    return ((list_nodet *)NULL);
+  else
+  {
+    list_nodet *result = malloc(16);
```
this should be changed to `sizeof(struct list)` so it doesn't randomly break in the future

NlightNFotis replied:
Agreed that this should be changed to something so that at least the value is self-documented.

xbauch replied:
The bug only appears if I don't use `sizeof`.

smowton replied:
Because `malloc(sizeof(type))` is special-cased to produce a `type*` instead of a `char*`

hannes-steffenhagen-diffblue reviewed (approved):
I don't quite get this, would prefer if someone more familiar with the code would sanity check this before merging

NlightNFotis reviewed (approved):
LGTM

petr-bauch committed:
Fix assigning to dynamically allocated structs

that is when

lhs :: [char]
rhs :: byte-extract-expr(struct, 0)

before we created a member expression at the offset and, in this case, assigned
to the first member of the struct. Now we assign the whole struct.

A test is included that was previously failing in the unwinding assertion (CBMC
did not see that the next pointer is eventually NULL).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0ed6adc34cff70937a074d73984ff45a0c855c8b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/128901538

smowton commented on src/pointer-analysis/value_set.cpp:
```diff
@@ -936,6 +936,11 @@ void value_sett::get_value_set_rec(
     const typet &op_type = ns.follow(byte_extract_expr.op().type());
     if(op1_offset.has_value() && op_type.id() == ID_struct)
     {
+      if(*op1_offset == 0)
```
This is forcing the not-found case when the access offset is zero, regardless of the access size. I can see no reason why zero is special here.

smowton reviewed (commented):
The problem is clearly genuine, but the solution seems like a somewhat-arbitrary hack. Why is the access against the zeroth member incorrect? Where did the write end up recorded in the value-set, and why was that write not similarly directed into a struct field as the read code is trying to do?

tautschnig commented:
Closing in favour of #6107: I've re-used the regression test (modulo a slight simplification and cleanup), but provided an implementation that should be less arbitrary.
